x/btcstaking/keeper/power_dist_change.go

```diff
@@ -56,31 +56,61 @@ func (k Keeper) UpdatePowerDist(ctx context.Context) {
 	// to construct the new distribution
 	newDc := k.ProcessAllPowerDistUpdateEvents(ctx, dc, events, maxActiveFps)
 
-	// find newly bonded finality providers and execute the hooks
-	newBondedFinalityProviders := newDc.FindNewActiveFinalityProviders(dc)
-	for _, fp := range newBondedFinalityProviders {
-		if err := k.hooks.AfterFinalityProviderActivated(ctx, fp.BtcPk); err != nil {
-			panic(fmt.Errorf("failed to execute after finality provider %s bonded", fp.BtcPk.MarshalHex()))
-		}
-	}
-
 	// record voting power and cache for this height
-	k.recordVotingPowerAndCache(ctx, newDc)
+	k.recordVotingPowerAndCache(ctx, dc, newDc, maxActiveFps)
 	// record metrics
 	k.recordMetrics(newDc)
 }
 
-func (k Keeper) recordVotingPowerAndCache(ctx context.Context, dc *types.VotingPowerDistCache) {
+func (k Keeper) recordVotingPowerAndCache(ctx context.Context, prevDc, newDc *types.VotingPowerDistCache, maxActiveFps uint32) {
+	if prevDc == nil {
+		panic("the previous voting power distribution cache cannot be nil")
+	}
+
+	// deep copy the previous dist cache if the new dist cache is nil
+	if newDc == nil {
```

Comment:
I think it would be cleaner to move this copying process to above the call to `recordVotingPowerAndCache` in case when there is no new events i.e somewehre under `// map everything in prev height to this height` comment.

Otherwise this `nil` check just generate questions like:
- oh why we can pass nil here ?
- should we also check `prevDc` to be non-nil ?
etc.

Where if this is handled on call site we can add comments along the line `that becouse there is no new power changing events, the new dc is just copy of the old one and the deep copy is need due to  recordVotingPowerAndCache modifying provided arguments.` (or something like that)

Reply:
Agreed allowing `nil` parameter is hacky. Fixed in [f3099ee](https://github.com/babylonlabs-io/babylon/pull/38/commits/f3099eeca7527d8ab1da471c2f177dee5ab40145)